### PR TITLE
Use RegDeleteValueW to delete registry value.

### PR DESCRIPTION
**Pull request by mcquin**

### CODE REVIEW

**Comment by chef-supermarket** [comment]
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.getchef.com/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## Non-GitHub Verified Committers

There are 1 commit author(s) whose commits are authored by a non-GitHub verified email address. Chef will have to manually verify that they are authorized to contribute.

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

**Comment by chefsalim** [comment]
I think we need to explicitly check the return value from RegDeleteValue, as a non-zero return value can actually be a failure.  Iff result == 0 then it is success.

File: lib/chef/win32/registry.rb [hunk 1]
```diff
@@ -91,8 +91,15 @@ def delete_value(key_path, value)
...
-            Chef::Log.debug("Deleted value #{value[:name]} from registry key #{key_path}")
+            # ::Win32::Registry#delete_value is broken in Ruby 2.1 (up to Ruby 2.1.6p336).
+            # This should be resolved a later release (see note #9 in link below).
+            # https://bugs.ruby-lang.org/issues/10820
+            if RegDeleteValueW(reg.hkey, wstring(value[:name]))
+              Chef::Log.debug("Deleted value #{value[:name]} from registry key #{key_path}")
+            else
+              Chef::Log.debug("Unable to delete value #{value[:name]} from registry key #{key_path}")
+              return false
+            end
```
Comment (mcquin) [inline lib/chef/win32/registry.rb]: @chefsalim (re: https://github.com/chef/chef/pull/3849#issuecomment-136494592) isn't that what's happening here? 0 is truthy, non-zero is falsey.

File: lib/chef/win32/registry.rb [hunk 2]
```diff
@@ -91,8 +91,15 @@ def delete_value(key_path, value)
             return true
           end
           hive.open(key, ::Win32::Registry::KEY_SET_VALUE | registry_system_architecture) do |reg|
-            reg.delete_value(value[:name])
-            Chef::Log.debug("Deleted value #{value[:name]} from registry key #{key_path}")
+            # ::Win32::Registry#delete_value is broken in Ruby 2.1 (up to Ruby 2.1.6p336).
+            # This should be resolved a later release (see note #9 in link below).
+            # https://bugs.ruby-lang.org/issues/10820
+            if RegDeleteValueW(reg.hkey, wstring(value[:name]))
```
Comment (jaym) [inline lib/chef/win32/registry.rb]: the return types for this block are inconsistent

Comment (mcquin) [inline lib/chef/win32/registry.rb]: Can you be more specific, please?

Comment (jaym) [inline lib/chef/win32/registry.rb]: one branch is returning whatever Chef::Log.debug returns, the other is returning a boolean

Comment (mcquin) [inline lib/chef/win32/registry.rb]: The non-boolean returning branch ultimately returns true. I can clean it up, though.

Comment (mcquin) [inline lib/chef/win32/registry.rb]: the first branch ultimately returns `true`.

**Comment by mcquin** [comment]
@chefsalim checking the exit status [here](https://github.com/chef/chef/blob/mcquin/reg-delete-value-ffi/lib/chef/win32/registry.rb#L97)

**Comment by chef-supermarket** [comment]
Hi. Your friendly Curry bot here. Just letting you know that all commit authors have become authorized to contribute. I have added the "Signed CLA" label to this issue so it can easily be found in the future.

**Comment by chefsalim** [comment]
Status check looks good to me. 

Not sure the unit test should directly be depending on the RegDeleteValue call though...

+      expect(@registry).to receive(:RegDeleteValueW).with(@reg_mock.hkey, value1[:name]).and_return(true)

**Comment by mcquin** [comment]
@chefsalim I'm not sure how else to stub it. Do you have any ideas?

**Comment by chefsalim** [comment]
Yeah, I see how it is tricky to eliminate the RegDeleteValue dependency entirely... another possibility could be to keep the RegDeleteValue message, but remove the to_wstring message expectation and the hKey definition in the reg double, and just do the following:

expect(@registry).to receive(:RegDeleteValueW).with(0, wstring("one")).and_return(true)

**Comment by mcquin** [comment]
Blocked by https://github.com/chef/chef/pull/3855

**Comment by mcquin** [comment]
Closing in favor of https://github.com/chef/chef/pull/3870

**Pull request closed without merging**